Algorithm.CSharp/FutureMarketOpenConsolidatorRegressionAlgorithm.cs

```diff
@@ -0,0 +1,147 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using QuantConnect.Interfaces;
+using QuantConnect.Data;
+using QuantConnect.Data.Consolidators;
+
+namespace QuantConnect.Algorithm.CSharp
+{
+    /// <summary>
+    /// Regression algorithm using a consolidator to check GetNextMarketClose() and GetNextMarketOpen()
+    /// are returning the correct market close and open times
+    /// </summary>
+    public class FutureMarketOpenConsolidatorRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
+    {
+        private Symbol _es;
+        private static List<DateTime> _expectedOpens = new List<DateTime>(){
+            new DateTime(2013, 10, 07, 16, 30, 0),
+            new DateTime(2013, 10, 09, 16, 30, 0),
+            new DateTime(2013, 10, 11, 16, 30, 0),
+            new DateTime(2013, 10, 15, 16, 30, 0),
+            new DateTime(2013, 10, 17, 16, 30, 0),
+            new DateTime(2013, 10, 19, 16, 30, 0)
+        };
+        private static List<DateTime> _expectedCloses = new List<DateTime>(){
+            new DateTime(2013, 10, 08, 16, 15, 0),
+            new DateTime(2013, 10, 10, 16, 15, 0),
+            new DateTime(2013, 10, 14, 16, 15, 0),
+            new DateTime(2013, 10, 16, 16, 15, 0),
+            new DateTime(2013, 10, 18, 16, 15, 0),
+            new DateTime(2013, 10, 20, 16, 15, 0)
+        };
+        private Queue<DateTime> _expectedOpensQueue = new Queue<DateTime>(_expectedOpens);
+        private Queue<DateTime> _expectedClosesQueue = new Queue<DateTime>(_expectedCloses);
+
+        public override void Initialize()
+        {
+            SetStartDate(2013, 10, 06);
+            SetEndDate(2013, 10, 14);
+
+            var es = AddSecurity(SecurityType.Future, "ES");
+
+            Consolidate<BaseData>(es.Symbol, time =>
+            {
+                var date = time;
+                if (time >= new DateTime(2013, 10, 21))
+                {
+                    date = Time;
+                }
```

Comment:
This is hiding a bug I think, see https://github.com/QuantConnect/Lean/blob/master/Common/Data/Consolidators/PeriodCountConsolidatorBase.cs#L385

---

On end of algorithm we should assert both queues are empty 👌